route-rs-packets/src/arp.rs

```diff
@@ -0,0 +1,275 @@
+use crate::{EthernetFrame, MacAddr, ARP_ETHER_TYPE};
+use std::convert::{TryFrom, TryInto};
+use std::net::IpAddr;
+
+pub enum ArpOp {
+    Request = 1,
+    Reply = 2,
+}
+
+pub enum ArpHardwareType {
+    Ethernet = 1,
+}
+
+const HARDWARE_TYPE_RANGE: (usize, usize) = (0, 2);
+const PROTOCOL_TYPE_RANGE: (usize, usize) = (2, 4);
+const HARDWARE_ADDR_LEN_RANGE: (usize, usize) = (4, 5);
+const PROTOCOL_ADDR_LEN_RANGE: (usize, usize) = (5, 6);
+const OPCODE_RANGE: (usize, usize) = (6, 8);
+
+///
+/// EthernetFrame wrapper with getters/setters for the packet structure described in RFC 826
+/// https://tools.ietf.org/html/rfc826
+///
+#[derive(Clone)]
+pub struct ArpFrame {
+    frame: EthernetFrame,
+}
+
+impl ArpFrame {
+    ///
+    /// Constructs a new, empty packet with a payload big enough for all ARP fields,
+    /// given some hardware/protocol address lengths.
+    ///
+    pub fn new(hardware_addr_len: u8, protocol_addr_len: u8) -> Self {
+        let payload_len = 8 + (2 * hardware_addr_len as usize) + (2 * protocol_addr_len as usize);
+        let payload: Vec<u8> = vec![0; payload_len];
+
+        let mut frame = EthernetFrame::empty();
+        frame.set_payload(payload.as_slice());
+
+        let mut arp_frame = ArpFrame { frame };
+        arp_frame.set_hardware_addr_len(hardware_addr_len);
+        arp_frame.set_protocol_addr_len(protocol_addr_len);
+        arp_frame
+    }
+
+    pub fn hardware_type(&self) -> u16 {
+        let (start, end) = HARDWARE_TYPE_RANGE;
+        u16::from_be_bytes(self.arp_data(start, end).try_into().unwrap())
+    }
+
+    pub fn protocol_type(&self) -> u16 {
+        let (start, end) = PROTOCOL_TYPE_RANGE;
+        u16::from_be_bytes(self.arp_data(start, end).try_into().unwrap())
+    }
+
+    pub fn hardware_addr_len(&self) -> u8 {
+        let (start, end) = HARDWARE_ADDR_LEN_RANGE;
+        u8::from_be_bytes(self.arp_data(start, end).try_into().unwrap())
+    }
+
+    pub fn protocol_addr_len(&self) -> u8 {
+        let (start, end) = PROTOCOL_ADDR_LEN_RANGE;
+        u8::from_be_bytes(self.arp_data(start, end).try_into().unwrap())
+    }
+
+    pub fn opcode(&self) -> u16 {
+        let (start, end) = OPCODE_RANGE;
+        u16::from_be_bytes(self.arp_data(start, end).try_into().unwrap())
+    }
+
+    pub fn sender_hardware_addr(&self) -> &[u8] {
+        let (start, end) = self.sender_hardware_addr_range();
+        self.arp_data(start, end)
+    }
+
+    pub fn sender_protocol_addr(&self) -> &[u8] {
+        let (start, end) = self.sender_protocol_addr_range();
+        self.arp_data(start, end)
+    }
+
+    pub fn target_hardware_addr(&self) -> &[u8] {
+        let (start, end) = self.target_hardware_addr_range();
+        self.arp_data(start, end)
+    }
+
+    pub fn target_protocol_addr(&self) -> &[u8] {
+        let (start, end) = self.target_protocol_addr_range();
+        self.arp_data(start, end)
+    }
+
+    pub fn set_hardware_type(&mut self, htype: u16) {
```

Comment:
Last change, this one is a new library thing, but trivial to add. 
Please make setter function return a `&mut Self` , this lets the user optionally chain setters. 
Like
A::new().set_1().set_2().set_3();

And the old uses are still valid since Rust will auto drop  returned references that are not assigned to a variable. See the other packets files for examples.